src/pages/CardAddPage.jsx

```diff
@@ -0,0 +1,297 @@
+import { useState, useMemo, useEffect } from 'react';
+import Head from '../components/Head';
+import Card from '../components/Card';
+import LabeledInput from '../components/LabeledInput';
+import InfoLabel from '../components/InfoLabel';
+import SubmitButton from '../components/SubmitButton';
+import validator from '../validation';
+import Input from '../components/Input';
+import { Page, CardSection, Form, FormRow, SubmitButtonContainer } from '../style/page';
+import {
+  cardNumberInputRegex,
+  expiredDateInputRegex,
+  ownerNameInputRegex,
+  numberRegex,
+} from '../constant/regularExpression';
+import { INPUT_TYPE } from '../constant';
+import MESSAGE from '../constant/message';
+import { CARD_NUMBER_MARK, CARD_NUMBER_SEPARATOR, DATE_SEPARATOR } from '../constant/mark';
+
+function CardAddPage() {
+  const [companyName, setCompanyName] = useState('포코카드');
+  const [cardNumbers, setCardNumbers] = useState([]);
+  const [expiredDate, setExpiredDate] = useState({ month: '', year: '' });
+  const [ownerName, setOwnerName] = useState('');
+  const [securityNumber, setSecurityNumber] = useState('');
+  const [password, setPassword] = useState(['', '']);
+  const [cardAddCondition, setCardAddCondition] = useState({
+    cardNumbers: false,
+    expiredDate: false,
+    ownerName: false,
+    securityNumber: false,
+    password: false,
+  });
+
+  const convertedCardNumbers = useMemo(() => {
+    return cardNumbers.map((cardNumber, index) =>
+      index >= 2 ? CARD_NUMBER_MARK.repeat(cardNumber.length) : cardNumber
+    );
+  }, [cardNumbers]);
+
+  const convertedExpiredDate = useMemo(() => {
+    return expiredDate.month || expiredDate.year
+      ? `${expiredDate.month}${expiredDate.month.length === 2 ? DATE_SEPARATOR : ''}${
+          expiredDate.year
+        }`
+      : '';
+  }, [expiredDate]);
+
+  useEffect(() => {
+    setCardAddCondition({
+      cardNumbers: validator.validateCardNumbers(cardNumbers.join(CARD_NUMBER_SEPARATOR)),
+      expiredDate: validator.validateExpiredDate(convertedExpiredDate),
+      ownerName: validator.validateOwnerName(ownerName),
+      securityNumber: validator.validateSecurityNumber(securityNumber),
+      password: validator.validatePassword(password.join('')),
+    });
+  });
```

Comment:
1. useEffect 의 두번째 인자에는 어떤 값이 들어가고, 그에 따라 렌더링 시에 호출이 얼마나 빈번히 되는지 알아볼까요?
2. state 로 저장하는 이유를 모르겠네요! 그냥 변수로 설정해도 될 것 같아요!

```suggestion
const isValidCardNumbers = validator.validateCardNumbers(cardNumbers.join(CARD_NUMBER_SEPARATOR));
const isValidExpiredDate = validator.validateExpiredDate(convertedExpiredDate);
const isValidOwnerName = validator.validateOwnerName(ownerName);
const isValidSecurityNumber = validator.validateSecurityNumber(securityNumber);
const isValidPassword = validator.validatePassword(password.join(''));
```

Reply:
1. useEffect의 첫 번째 인자에는 함수가 전달되고, 두 번째 인자에는 의존성 배열이 전달된다고 알고 있습니다. 렌더링 시 두 번째 인자로 전달한 배열의 값 중 변경된 값이 있으면 첫 번째 인자로 전달한 함수가 실행된다고 알고있습니다.
2. 동의합니다. 지금처럼 state로 저장을 하는 것이 적절하지 않은 거 같습니다.

```javascript
useEffect(() => {
    console.log('useEffect가 실행됨');
    setCardAddCondition({
      cardNumbers: validator.validateCardNumbers(cardNumbers.join(CARD_NUMBER_SEPARATOR)),
      expiredDate: validator.validateExpiredDate(convertedExpiredDate),
      ownerName: validator.validateOwnerName(ownerName),
      securityNumber: validator.validateSecurityNumber(securityNumber),
      password: validator.validatePassword(password.join('')),
    });
  });
```

언급 해주신 useEffect에 console.log()를 출력하게 수정확인한 결과 아래의 gif 처럼 log가 계속 찍히는 것을 확인했습니다.

![무한 렌더링](https://user-images.githubusercontent.com/52148907/166192392-5c295961-4c81-4251-9476-46c3eb06351b.gif)

추측이지만 useEffect 내부에 setState가 있어서 계속 렌더링이 되는 거 같습니다..! 따라서 해당 부분은 제안 해주신 대로 state 대신에 변수로 수정하겠습니다!

Reply:
무한루프가 될 수 밖에 없는 구조예요! 

`useEffect` -> `setCardAddCondition()` 실행 -> 상태바꼈으니 `useEffect` 또 실행 -> 또 `setCardAddCondition` 실행 -> ...

그래서 useEffect 내에서 조건 분기도 없이 / deps 배열도 없이 state 를 set 하는건 굉장히 위험한 일입니다!